### ISSUE TITLE
Please bring back .select() or make a way of retaining scraped page!

### ISSUE THREAD

**Issue by ssteinerx** [lapwinglabs/x-ray#48]
I used to use:
  var page = xray('http://url.com')

to get the page, then:

page.select([{
  thing: '.classname',
  stuff: '.other_classname'
}])...and so on;

This allowed me to get the page once, then scrape it in various ways.

Did I miss something, or did this wonderful capability go away?

If so, please bring it back!

S

**Comment by matthewmueller** [comment]
Hm, not really following, can't you just do:

```js
xray('http:url.com', [{
  thing: '.className',
  stuff: '.other_classname'
}])(fn)
```

`.select` didn't actually do anything until you ran it anyway.

**Comment by ssteinerx** [comment]
Yes, but I just want to snag the HTML once, then play with it in different ways.

I find the way JavaScript programmers tend to try to write everything into one humungous expression with embedded anonymous functions etc. incredibly difficult to parse and even harder to debug.

I tend to unroll things and approach each piece individually.

So, rather than try to parse every thing out in one shot, I tend to do related groups in individual, named functions.

Also, when you want to $root something somewhere, this is the only way I know of to do that.

Steve

**Comment by matthewmueller** [comment]
Can you give me an example of rolling this out? Im under the impression
that the composable API does a better job of doing that then the fluent one
right now.

**Comment by ssteinerx** [comment]
Since there's no documentation on the "composable API" nor the "fluent API", I have no basis for comparison and I'm not even convinced I know how you're using those terms.

**Comment by matthewmueller** [comment]
Okay I'm going to assume the tone of your message is not aggressive and
more confusion though its certainly hard to be sure.
The main thing I was asking for was an example of how you're using the old
API with select(). I'm fairly certain the new API can accommodate your
request, given that I wrote both.
Also please keep in mind this is free software that I write in my free
time. If you are unhappy with the changes feel free to use the older
version, fork it, or try a different scraping library entirely.

**Comment by ssteinerx** [comment]
I apologize, I understand how that might have come across.  I really appreciate your efforts in creating this and did not mean to sound ungrateful.  I think my frustration with having spent 2 (very) long days, and having just gotten most things working before upgrading to 2.0 came out may have leaked through a bit ;-).

Here's how I was using it.  

/* Read the URL once, then use it over and over */
var xrayed = xray('www.studleyflorist.com');

/* Snag all the sidenav links on the page, store them if we don't already have them */
xrayed.select([{
 link: '.sidenav_sub_indices.sidenav_arrowside_empty';
}])(function (err, arr){
  ...store in 'links' database table
});

/* Snag any products on the page, put them in the database */
xrayed.select([{
  title: '.promo_name',
  link: '.promo_img a@href',
  prod_code: '.prod-code',
  price: '.price',
  }])(function (err, arr) {
   ... Write 'promo' to database 'promo' table.
...}
});

/* Snag each image link, download the referenced files into the .images directory */
xrayed.select([{
  link: '.promo_img a@href'
}])(function (err, arr) {
  if (err) throw err;
  ... Download the images and write the files into the ./images directory...
}]);

In other words, I need post-processing of the results to do three different things.

While I might be able to construct a huge single call to xray(), that doesn't seem like the modular, debuggable thing to do and re-reading the URL for each operation seems really inefficient.

Thanks,

Steve

**Comment by matthewmueller** [comment]
Ah I see. So the operations of reading a URL is more or less a "no-op" for both the current API and the previous API, so I efficiency doesn't change there. The function `function(err, arr) { ... }` in both APIs will perform the actual request. In your example, x-ray will still perform a new request each time the function is called.

But to your point of modularity being able to break up the selectors. The new API reverses this a little bit, so that we can do some neat things with crawling. You can still separate the URL passing from the selection in the following way:

** select.js **

```js
module.exports = xray({
  title: '.promo_name',
  link: '.promo_img a@href',
  prod_code: '.prod-code',
  price: '.price'
})
```

** index.js **

```js
var select = require('./select.js')
select('www.studleyflorist.com', function(err, arr) {
  // ...
})
```

**Comment by ssteinerx** [comment]
Thanks, I didn't realize that each of the old 'select's would re-request the URL.

If I just snag the HTML (using built in request() or whatever) and pass it into each select()-ish function, does that buy me the efficiency I'm looking for?

Also, is $root still a valid concept?  In some cases, I want to only iterate over children of a particular #id.

Can I use that in my 'select()' param?

Again thanks for your help.  I do sometimes get tripped up by the semantics of the bizarre ways JavaScript is currently being abused...  I've been a programmer long enough to know that yesterday's 'clever' code is going to be an epic maintenance nightmare six months from now.

![image](https://cloud.githubusercontent.com/assets/43063/8026260/a04ce84e-0d3d-11e5-9a66-dace38672127.png)

Thanks,

S

**Comment by Kikobeats** [comment]
Can we close this?

**Comment by ssteinerx** [comment]
Closed, thanks!
